ledger/simulation/simulator_test.go

```diff
@@ -0,0 +1,665 @@
+
+	// Compile AVM program
+	ops, err := logic.AssembleString(trivialAVMProgram)
+	require.NoError(t, err, ops.Errors)
+	prog := ops.Program
+
+	// Create program and call it
+	futureAppID := 1
+	txgroup := []transactions.SignedTxn{
+		{
+			Txn: transactions.Transaction{
+				Type:   protocol.ApplicationCallTx,
+				Header: makeTxnHeader(sender),
+				ApplicationCallTxnFields: transactions.ApplicationCallTxnFields{
+					ApplicationID:     0,
+					ApprovalProgram:   prog,
+					ClearStateProgram: prog,
+					LocalStateSchema: basics.StateSchema{
+						NumUint:      0,
+						NumByteSlice: 0,
+					},
+					GlobalStateSchema: basics.StateSchema{
+						NumUint:      0,
+						NumByteSlice: 0,
+					},
+				},
+			},
+		},
+		{
+			Txn: transactions.Transaction{
+				Type:   protocol.ApplicationCallTx,
+				Header: makeTxnHeader(sender),
+				ApplicationCallTxnFields: transactions.ApplicationCallTxnFields{
+					ApplicationID: basics.AppIndex(futureAppID),
+				},
+			},
+		},
+	}
+
+	err = attachGroupID(txgroup)
+	require.NoError(t, err)
+
+	_, _, err = s.Simulate(txgroup)
+	require.NoError(t, err)
+}
+
+func TestRejectAppCall(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	l, accounts, makeTxnHeader := prepareSimulatorTest(t)
+	defer l.Close()
+	s := simulation.MakeSimulator(l)
+	sender := accounts[0].addr
+
+	// Compile AVM program
+	ops, err := logic.AssembleString(rejectAVMProgram)
+	require.NoError(t, err, ops.Errors)
+	prog := ops.Program
+
+	// Create program and call it
+	txgroup := []transactions.SignedTxn{
+		{
+			Txn: transactions.Transaction{
+				Type:   protocol.ApplicationCallTx,
+				Header: makeTxnHeader(sender),
+				ApplicationCallTxnFields: transactions.ApplicationCallTxnFields{
+					ApplicationID:     0,
+					ApprovalProgram:   prog,
+					ClearStateProgram: prog,
+					LocalStateSchema: basics.StateSchema{
+						NumUint:      0,
+						NumByteSlice: 0,
+					},
+					GlobalStateSchema: basics.StateSchema{
+						NumUint:      0,
+						NumByteSlice: 0,
+					},
+				},
+			},
+		},
+	}
+
+	err = attachGroupID(txgroup)
+	require.NoError(t, err)
+
+	_, _, err = s.Simulate(txgroup)
+	require.ErrorAs(t, err, &simulation.EvalFailureError{})
+	require.ErrorContains(t, err, "transaction rejected by ApprovalProgram")
+}
+
+func TestSignatureCheck(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	l, accounts, makeTxnHeader := prepareSimulatorTest(t)
+	defer l.Close()
+	s := simulation.MakeSimulator(l)
+	sender := accounts[0].addr
+
+	txgroup := []transactions.SignedTxn{
+		{
+			Txn: transactions.Transaction{
+				Type:   protocol.PaymentTx,
+				Header: makeTxnHeader(sender),
+				PaymentTxnFields: transactions.PaymentTxnFields{
+					Receiver: sender,
+					Amount:   basics.MicroAlgos{Raw: 0},
+				},
+			},
+		},
+	}
+
+	// should error without a signature
+	_, missingSignatures, err := s.Simulate(txgroup)
+	require.NoError(t, err)
+	require.True(t, missingSignatures)
```

Comment:
nit: "should error" is a bit misleading, considering we expect `err` to be nil here